### ISSUE TITLE
Support Plug-in for native SDK intergration

### ISSUE THREAD

**Issue by iwae** [cocos/cocos-engine#14250]
### Use Case

In other engines, there is a "Plug-ins" folder, which supports includes the scripts created outside the engine in the form of a plug-in, which will be a great improvement for native cases.
https://docs.unity3d.com/Manual/Plugins.html

### Problem Description

For example, we can simply include the wwise SDKs for ios and android in the Plug-ins folder, and also a typescript to call the JSB bridge for them, so user could simply just call the typescript within the TS scripts. The editor will deals with the SDK into the workplace of the native projects and register the permissions for the SDKs.

### Proposed Solution

Like we have a Add-ons folder that contains 4 folders with the names " IOS", "Android","Mac","Win", which is pre-created for these 4 native platforms, each sub-folder cantains a "libs" folder for native SDK, a minefist file would be requested for registers, and a "SRC" folder for TypeScript files.

### How it works

When the engine build projects for each native platform, it will include the SDK from libs for the target platform, and also register from the minefist file, so the user will only have to call the TS files for JSB .

### Alternatives Considered

It will be a good solutions for both common SDKs and AD/Monetization SDKs.

**Comment by minggo** [comment]
I think can just use current add-on mechanism to support native SDK.

**Comment by jareguo** [comment]
## EN

The existing CMake solution is not friendly enough for SDK developers, and due to the changes in the engine project structure, it is difficult to ensure compatibility after adaptation. Often, old versions of the SDK cannot be used in new versions.

On the one hand, this is due to the C/C++ plugin support of the engine (similar to the needs of **), and on the other hand, it is achieved by using the editor to automatically integrate native projects. To achieve this, you can refer to Unity, where plugins can be directly imported into the editor and divided into Managed and [Native](https://docs.unity3d.com/Manual/NativePlugins.html). The former is equivalent to our current plugin script, and the latter is not yet supported.

![image](https://user-images.githubusercontent.com/1503156/228199534-e3c9402b-427c-4efa-b4f7-315c9cacf863.png)

See https://docs.unity3d.com/2023.2/Documentation/Manual/PluginInspector.html

It is necessary to design a similar mechanism, involving:

1. The engine's own native plugin system (native plugin) for expanding initialization, rendering, platform API calls, Shader compilation and other processes.
1.1 The JSB binding mechanism of native plugins, which can inject binding code into V8.
1.2 The compilation mechanism of native plugins (using native IDEs on various platforms).
2. The editor's support for importing native plugins, which involves integration with the editor's extension mechanism, native plugin management, configuration, JS binding interface compilation, automatic generation to CMakeFile during construction, and other processes.
3. The editor's support for native plugins in native scenes, which should be able to load native plugins in native scenes (otherwise the effect cannot be previewed).

Reference:
https://developers.google.com/admob/unity/quick-start
https://www.shuzhiduo.com/A/MAzAjMDoJ9/

## ZH

现有的 CMake 方案对 SDK 开发者还是不够友好，而且由于引擎项目结构的改动，适配后的兼容性很难保证。往往老版本的 SDK 无法用于新版本。
这里面一方面是引擎的 C/C++ 插件支持（类似**汽车的需求），另一方面是借助编辑器来做到原生工程的自动集成，要做到这一点，可以参考 Unity，在编辑器中可以直接导入插件，分成 Managed 和 [Native](https://docs.unity3d.com/Manual/NativePlugins.html) 两种。前者相当于我们现在的插件脚本，后者我们现在还不支持。
![image](https://user-images.githubusercontent.com/1503156/228199534-e3c9402b-427c-4efa-b4f7-315c9cacf863.png)
See https://docs.unity3d.com/2023.2/Documentation/Manual/PluginInspector.html

需要设计一个类似的机制，涉及到
1. 引擎自身原生的插件系统（原生插件），用于扩展初始化、渲染、平台 API 调用、Shader 编译等环节
1.1 原生插件的 JSB 绑定机制，能够注入绑定代码到 V8
1.2 原生插件的编译机制（借助各平台原生 IDE）
2. 编辑器对原生插件的导入支持，涉及到与编辑器扩展机制的整合、原生插件管理、配置、JS 绑定接口编译、构建时自动生成到 CMakeFile 等环节
3. 编辑器原生场景对原生插件的支持，需要能在原生化的场景中加载原生插件（否则预览不了效果）

参考：
https://developers.google.com/admob/unity/quick-start 
https://www.shuzhiduo.com/A/MAzAjMDoJ9/

**Comment by jareguo** [comment]
In addition, there is another question: is it sufficient to use this plugin mechanism for integrating common third-party SDKs?
For example, for the Java SDK, if we write a native plugin that calls Java through JNI and bind it to TS script, we can directly call Java in the project using TS.

**Comment by iwae** [comment]
It should be OK for most cases, other engines  support Call Java/ AAR

![image](https://user-images.githubusercontent.com/26038745/228208691-15451a89-9053-4b73-8acb-b4ed67a87fc2.png)

**Comment by jareguo** [comment]
Yes, we can go further and automatically generate binding code for jar/aar

**Comment by ankye** [comment]
这个有什进展吗?目前我跑通了native c++的插件逻辑,但是不知道怎么将java和oc代码注入插件目录,可能还包含native的原生ui
